src/main/java/org/springframework/data/elasticsearch/core/ElasticsearchEntityMapper.java

```diff
@@ -0,0 +1,676 @@
+				}
+			}
+		}
+
+		return (R) target;
+	}
+
+	protected Object readSimpleValue(@Nullable Object value, TypeInformation<?> targetType) {
+
+		if (value == null) {
+			return null;
+		}
+
+		if (ClassTypeInformation.OBJECT.equals(targetType)
+				|| (targetType != null && value.getClass().equals(targetType.getActualType()))) {
+			return value;
+		}
+
+		if (conversionService.canConvert(value.getClass(), targetType.getType())) {
+			return conversionService.convert(value, targetType.getType());
+		}
+
+		throw new MappingException(
+				String.format("Unable to map %s of type %s to %s", value, value.getClass(), targetType.getType()));
+	}
+
+	// --> WRITE
+
+	@Override
+	public Map<String, Object> mapObject(Object source) {
+
+		LinkedHashMap<String, Object> target = new LinkedHashMap<>();
+		write(source, target);
+		return target;
+	}
+
+	@Override
+	public void write(@Nullable Object source, Map sink) {
+
+		if (source == null) {
+			return;
+		}
+
+		if (source instanceof Map) {
+
+			sink.putAll((Map) source);
+			return;
+		}
+
+		Class<?> entityType = ClassUtils.getUserClass(source.getClass());
+		TypeInformation<? extends Object> type = ClassTypeInformation.from(entityType);
+
+		if (requiresTypeHint(type, null)) {
+			typeMapper.writeType(type, sink);
+		}
+
+		doWrite(source, sink, type);
+	}
+
+	protected void doWrite(@Nullable Object source, Map<String, Object> sink,
+			@Nullable TypeInformation<? extends Object> typeHint) {
+
+		if (source == null) {
+			return;
+		}
+
+		Class<?> entityType = source.getClass();
+		Optional<Class<?>> customTarget = conversions.getCustomWriteTarget(entityType, Map.class);
+
+		if (customTarget.isPresent()) {
+
+			sink.putAll(conversionService.convert(source, Map.class));
+			return;
+		}
+
+		if (typeHint != null) {
+
+			ElasticsearchPersistentEntity<?> entity = mappingContext.getRequiredPersistentEntity(typeHint);
+			writeEntity(entity, source, sink, null);
+			return;
+		}
+
+		// write Entity
+		ElasticsearchPersistentEntity<?> entity = mappingContext.getRequiredPersistentEntity(entityType);
+		writeEntity(entity, source, sink, null);
+	}
+
+	protected void writeEntity(ElasticsearchPersistentEntity<?> entity, Object source, Map<String, Object> sink,
+			@Nullable TypeInformation containingStructure) {
+
+		PersistentPropertyAccessor<?> accessor = entity.getPropertyAccessor(source);
+
+		if (requiresTypeHint(entity.getTypeInformation(), containingStructure)) {
+			typeMapper.writeType(entity.getTypeInformation(), sink);
+		}
+		writeProperties(entity, accessor, sink);
+	}
+
+	protected void writeProperties(ElasticsearchPersistentEntity<?> entity, PersistentPropertyAccessor<?> accessor,
+			Map<String, Object> sink) {
+
+		for (ElasticsearchPersistentProperty property : entity) {
+
+			if (!property.isWritable()) {
+				continue;
+			}
+
+			Object value = accessor.getProperty(property);
+
+			if (value == null) {
+				continue;
+			}
+
+			if (!conversions.isSimpleType(value.getClass())) {
```

Comment:
We have quite a few `isSimpleType(…)` calls. Do we want to consolidate these into methods?